### ISSUE TITLE
set random seed for batch job

### ISSUE THREAD

**Issue by mheldman** [readdy/readdy#227]
Hi,

I am using Readdy for a reversible A + B <-> C simulation. I want to compute some statistics to high accuracy and so I am running a lot of trials, which would best be done as a batch job. Is there a way to set the seed for the random number generator for each job?

Sincerely,

Max Heldman

**Comment by clonker** [comment]
Hi Max,

unfortunately it is currently not possible to set the seed manually, but it would a nice feature to have indeed. Part of the problem is that random generators are not thread-safe (because they are stateful) so we instantiate new ones per thread which are all randomly seeded, otherwise you'd get artifacts. I could imagine setting the seed for the singlecpu (single-threaded) implementation, though. That one is much better suited for sampling anyways.
Right now I don't have the capacity to implement this but I'll put it on my to-do list. Or you can have a stab at implementing it under my guidance.

Cheers,
Moritz

**Comment by mheldman** [comment]
Hi Moritz,

Yes, I am running in single cpu mode- I'm using fully reversible reaction dynamics which it seems does not have a multithread implementation. 

I'm happy to try to implement it myself with guidance. I would have tried already but I am not that comfortable with C++ so the layout of the source code is a little complicated for me. 

Max

**Comment by clonker** [comment]
The first step would be implementing an option to the `KernelConfiguration` of the single cpu kernel (see [here](https://github.com/readdy/readdy/blob/d426708ed146655b90583980f1a7a7af76c49983/include/readdy/api/KernelConfiguration.h#L148)). You can take inspiration on how to implement that based on the other kernels and their configurations. A random generator (as in [here](https://github.com/readdy/readdy/blob/d426708ed146655b90583980f1a7a7af76c49983/include/readdy/model/RandomProvider.h#L61)) should be saved in the [SCPUStateModel](https://github.com/readdy/readdy/blob/master/include/readdy/kernel/singlecpu/SCPUStateModel.h). We can worry about the places where the generator has to be used and the python binding afterwards. :)

**Comment by mheldman** [comment]
Ok, I think for now I am going to use a workaround by setting the seed in an environment variable and modifying the randomlySeededGenerator function to read it's value. 

By the way, I think readdy is not compatible with the newest version of catch2 right now. In version 3 they changed the name of their header file. For the conda build script to work I had to set catch2=2 in meta.yaml.

Max

**Comment by clonker** [comment]
Hi, the way you pasted it here it is (unless im missing something) still not deterministic. Only if the `std::seed_seq` is completely initialized from your environment variable you will have quasireproducible results. Thanks for letting me know about the new catch2 version, will fix asap. :)

**Comment by mheldman** [comment]
Yep, it's not deterministic as written but as you suggest it can be modified to be so (maybe replace the r() and clck with zeros).
